spring-test/src/main/java/org/springframework/test/web/client/match/MultipartFormDataRequestMatchers.java

```diff
@@ -0,0 +1,252 @@
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.test.web.client.match;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import org.apache.commons.io.IOUtils;
+import org.hamcrest.Matcher;
+import org.hamcrest.Matchers;
+import org.jetbrains.annotations.NotNull;
+
+import org.springframework.core.io.Resource;
+import org.springframework.http.MediaType;
+import org.springframework.http.client.ClientHttpRequest;
+import org.springframework.mock.http.client.MockClientHttpRequest;
+import org.springframework.mock.web.MockHttpServletRequest;
+import org.springframework.test.web.client.RequestMatcher;
+import org.springframework.util.Assert;
+import org.springframework.util.MultiValueMap;
+import org.springframework.web.multipart.MultipartFile;
+import org.springframework.web.multipart.MultipartHttpServletRequest;
+import org.springframework.web.multipart.commons.CommonsMultipartResolver;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.springframework.test.util.AssertionErrors.assertEquals;
+import static org.springframework.test.util.AssertionErrors.fail;
+
+/**
+ * Factory for assertions on the multipart form data parameters. Handles only {@link MediaType#MULTIPART_FORM_DATA}
+ *
+ * <p>An instance of this class is typically accessed via {@link ContentRequestMatchers#multipart()}
+ *
+ * @author Valentin Spac
+ * @since 5.2
+ */
+public class MultipartFormDataRequestMatchers {
+
+	public RequestMatcher param(String parameter, String... expectedValues) {
+		List<Matcher<? super String>> matcherList = Arrays.stream(expectedValues)
+				.map(Matchers::equalTo)
+				.collect(Collectors.toList());
+
+		return this.param(parameter, matcherList);
+	}
+
+	@SafeVarargs
+	@SuppressWarnings("varargs")
+	public final RequestMatcher param(String parameter, Matcher<? super String>... matchers) {
+		return this.param(parameter, Arrays.stream(matchers).collect(Collectors.toList()));
+	}
+
+	public RequestMatcher param(String parameter, Matcher<Iterable<? extends String>> matchers) {
+		return new AbstractFormDataRequestMatcher() {
+			@Override
+			protected void matchInternal(MultipartHttpServletRequest request) {
+				Map<String, String[]> requestParams = request.getParameterMap();
+				assertValueCount(parameter, requestParams, 1);
+
+				String[] values = requestParams.get(parameter);
+				assertThat("Request parameter [" + parameter + "]", Arrays.asList(values), matchers);
+			}
+		};
```

Comment:
This might look better if the abstract class is replaced with a lambda that calls a static method to parse the multipart content.

Reply:
Agree. Please see latest changes